### PR TITLE
Detect mistyped associated consts in `Instance::resolve`.

### LINKED ISSUE #70942: [trait constants] thread 'rustc' panicked at 'assertion failed: `(left == right)` ...'
I'm seeing an internal compiler error on the following input (found by [fuzz-rustc](https://github.com/dwrensha/fuzz-rustc)):

```rust
trait Nat {
    const VALUE: usize;
}

struct Zero;

impl Nat for Zero {
    const VALUE: i32 = 0;
}

fn main() {
    let _: [i32; Zero::VALUE] = [];
}
```

```
$ rustc main.rs
error[E0326]: implemented const `VALUE` has an incompatible type for trait
 --> main.rs:8:18
  |
2 |     const VALUE: usize;
  |                  ----- type in trait
...
8 |     const VALUE: i32 = 0;
  |                  ^^^ expected `usize`, found `i32`

thread 'rustc' panicked at 'assertion failed: `(left == right)`
  left: `8`,
 right: `4`', /rustc/485c5fb6e1bf12cd11a8fac5ee94962e17cff74b/src/libstd/macros.rs:16:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.44.0-nightly (485c5fb6e 2020-04-08) running on x86_64-unknown-linux-gnu

error: aborting due to previous error

For more information about this error, try `rustc --explain E0326`.
```

### PR DESCRIPTION
**Pull request by eddyb**
*Based on #71049 to prevent redundant/misleading downstream errors.*

Fixes #70942 by refusing to resolve an associated `const` if it doesn't have the same type in the `impl` that it does in the `trait` (which we assume had errored, and `delay_span_bug` guards against bugs).

### CODE REVIEW

**Comment by eddyb** [comment]
Huh, I guess this is fast enough.

r? @oli-obk cc @nikomatsakis

**Comment by RalfJung** [comment]
Thanks! I take it we still need to keep the [`TransmuteSizeDiff` hack](https://github.com/rust-lang/rust/blob/96d77f0e5f103612d62b85938aacfb33f5768433/src/librustc_mir/interpret/place.rs#L942)?

File: src/test/ui/consts/issue-70942-trait-vs-impl-mismatch.stderr [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+error[E0308]: mismatched types
+  --> $DIR/issue-70942-trait-vs-impl-mismatch.rs:13:33
+   |
+LL |     let _: [i32; Zero::VALUE] = [];
+   |            ------------------   ^^ expected `Zero::VALUE`, found `0usize`
+   |            |
+   |            expected due to this
+   |
+   = note: expected array `[i32; _]`
+              found array `[_; 0]`
```
Comment (eddyb) [inline src/test/ui/consts/issue-70942-trait-vs-impl-mismatch.stderr:19]: I don't like this error, but I'm not sure how to resolve it, unless... oh yeah I know what to do.

Comment (eddyb) [reply src/test/ui/consts/issue-70942-trait-vs-impl-mismatch.stderr:19]: The fix turned out to be a bigger refactor than anticipated, so I  opened a separate PR: #71049.

**Comment by eddyb** [comment]
Blocked on #71049.

**Review by nikomatsakis — APPROVED** [review]
(All looks good to me, from a quick read)

**Pull request merged**